src/coreclr/nativeaot/System.Private.CoreLib/src/System/Reflection/Assembly.NativeAot.cs

```diff
@@ -16,12 +18,42 @@ public abstract partial class Assembly : ICustomAttributeProvider, ISerializable
         [System.Runtime.CompilerServices.Intrinsic]
         public static Assembly GetExecutingAssembly() { throw NotImplemented.ByDesign; } //Implemented by toolchain.
 
+        [DynamicSecurityMethod]
         public static Assembly GetCallingAssembly()
         {
             if (AppContext.TryGetSwitch("Switch.System.Reflection.Assembly.SimulatedCallingAssembly", out bool isSimulated) && isSimulated)
                 return GetEntryAssembly();
 
-            throw new PlatformNotSupportedException();
+            if (!StackTrace.IsSupported)
+                throw new NotSupportedException(SR.NotSupported_StackTraceSupportDisabled);
+
+            // We want to be able to handle GetCallingAssembly being called from Main (CoreCLR returns
+            // the assembly of Main), and also GetCallingAssembly being called from things like
+            // delegate invoke thunks. We do this by making the the definition of "calling assembly"
+            // a bit more loose.
+
+            // Technically we want skipFrames: 2 since we're interested in the frame that
+            // called the method that calls GetCallingAssembly, but we might need the first frame
+            // later in this method.
+            var stackTrace = new StackTrace(skipFrames: 1);
```

Comment:
This looks expensive. What's the performance of this implementation of `GetCallingAssembly` compared to regular CoreCLR in typical case (e.g. `GetCallingAssembly` called somewhere from ASP.NET)?

Reply:
Is your concern that we obtain entire stack trace as opposed to just a single frame? Both of these end up doing similar work anyway (they call RhGetCurrentStackTrace), just one of them will throw out everything but the one frame.

We could introduce a new specialized stack walking API, but I'm only doing this because it's cheap to implement.

I'm also fine just won't fixing the bug and telling people they can't do this until https://github.com/dotnet/csharplang/issues/4984 is implemented because once this is no longer cheap to implement, it's not worth the effort.

Reply:
My concern was that this is replacing naot functionality issue with naot-specific performance issue. The performance of this implementation can be much worse compared to CoreCLR one.

introducing specialized stackwalking method that does not walk the whole stack and some perf numbers would address my concern.

I am fine with won’t fixing.

Reply:
> introducing specialized stackwalking method that does not walk the whole stack and some perf numbers would address my concern.

Right, that would double the amount of work and risk/bug potential, tipping this over to the "not worth the effort given the limited benefit" category. Very few people missed this on .NET Native, and so do on native AOT. The csharplang issue is a much better fix. The only reason why I even looked at this is not that _some_ users would benefit from this, but that we produce no warning about potential behavior difference, but this is a rarely used API and I can live with that hole.